### PR TITLE
MINOR: A few cleanups of usage of Either in TC

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (hachikuji) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Note this was a bug spotted by @apurvam

**Comment by guozhangwang** [comment]
@hachikuji are the Jenkins failures relevant?

**Comment by hachikuji** [comment]
@guozhangwang Yes they are. I pushed a fix.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Can you please specify the result type of the public `getTransactionState` method as well?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Would it work to move the creation of this instance higher up and to just do the txn metadata within these two `case` statements?

**Review by ijuma — APPROVED** [review]
Thanks for the PR, this is a good improvement, it makes the code more idiomatic and avoids some bad patterns like matches that are not exhaustive. LGTM. Just one question and comment. No re-review needed.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Probably not what you had in mind, but I pushed a simplification.

**Comment by hachikuji** [comment]
By the way, I'm going to hold off merging this until #3287 is in.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (ijuma) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: You can simplify this one in a similar way as well:

```scala
Option(cacheEntry.metadataPerTransactionalId.putIfNotExists(transactionalId, txnMetadata)).getOrElse {
  txnMetadata
}
```

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (ijuma) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Your simplification looks good. It's roughly what I had in mind although I hadn't worked out all the details. I suggested one additional simplification.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala
Comment (guozhangwang) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: nit: `Expected non-empty transaction metadata returned` reads a bit unclear. Were you trying to say `Unexpected empty transaction metadata returned while putting $txnMetadata`?

Comment (hachikuji) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: Ack

**Comment by guozhangwang** [comment]
One minor comment, otherwise LGTM!

**Comment by hachikuji** [comment]
Rebased against trunk. I will merge after the builds finish.

**Pull request closed without merging**
